File: geekbang/algorithm/source-code/2/lc111 二叉树的最小深度.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDepth = function(root) {
    // 不同于maxDepth的另一种写法
    // 也可以采用maxDepth的写法
    var depth = 0;
    var ans = 1 << 30;

    var recur = function(node) {
        if (node == null) {
            ans = Math.min(ans, depth);
            return;
        }
        if (node.left == null && node.right == null) {
            ans = Math.min(ans, depth + 1);
            return;
        }
        depth++;
        if (node.left != null) recur(node.left);
        if (node.right != null) recur(node.right);
        depth--;
    }
    
    recur(root);
    return ans;
};

File: geekbang/algorithm/source-code/2/lc111 二叉树的最小深度.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        // 不同于maxDepth的另一种写法
        // 也可以采用maxDepth的写法
        recur(root);
        return ans;
    }
private:
    void recur(TreeNode* node) {
        if (node == nullptr) {
            ans = min(ans, depth);
            return;
        }
        if (!node->left && !node->right) {
            ans = min(ans, depth + 1);
            return;
        }
        depth++;
        if (node->left) recur(node->left);
        if (node->right) recur(node->right);
        depth--;
    }

    int depth = 0;
    int ans = 1 << 30;
};